experimental/ecma402_listformat/src/lib.rs

```diff
@@ -0,0 +1,565 @@
+            /// Selects a [Type] for the formatted list.
+            pub in_type: Type,
+        }
+    }
+
+    /// A factory object containing the methods used to construct new [Format], as well
+    /// as associated functions prescribed by `Intl.ListFormat`.
+    ///
+    /// While ECMA 402 originally has functions under `Intl`, we probably want to
+    /// obtain a separate factory from each implementor.
+    pub trait Factory<'a> {
+        /// A better name for this associated type is most welcome.
+        type Ret: Format<'a>;
+
+        /// The type of error reported, if any.
+        type Error: std::error::Error;
+
+        /// Creates a new [Format].
+        ///
+        /// Creation may fail, for example, if the locale-specific data is not loaded, or if
+        /// the supplied options are inconsistent.
+        fn try_new(l: impl crate::Locale<'a>, opts: options::In) -> Result<Self::Ret, Self::Error>;
+
+        /// The type used to implement [crate::Locale].
+        type Locale: crate::Locale<'a> + 'a;
+
+        /// The type of iterator returned by the call to `supported_locales_of`.
+        type LocaleIter: std::iter::Iterator<Item = &'a Self::Locale>;
+
+        fn supported_locales_of<I>(l: I) -> Result<Self::LocaleIter, Self::Error>
+        where
+            I: IntoIterator<Item = &'a Self::Locale>;
+    }
+
+    use std::fmt;
+
+    /// The package workhorse: formats supplied pieces of text into an ergonomically formatted
+    /// list.
+    ///
+    /// Use the appropriate [Factory] to create new instances of [Format].
+    pub trait Format<'a> {
+        /// Formats `list` into the supplied standard `formatter`.
+        ///
+        /// The original [ECMA 402 function][ecma402fmt] returns a string.  This is likely the only
+        /// reasonably generic option in JavaScript so it is adequate.  In Rust, however, it is
+        /// possible to pass in a standard formatting strategy (through `formatter`).
+        ///
+        ///   [ecma402fmt]:
+        ///   https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat/format
+        ///
+        /// This makes it unnecessary for [Format] to implement its own, and can
+        /// completely avoid constructing any intermediary representation.  This, in turn,
+        /// allows the user to provide a purpose built formatter, or an "exotic" one if needed.
+        ///
+        /// A purpose built formatter could be one that formats into a fixed-size buffer; or
+        /// another that knows how to format strings into a DOM.  If ECMA 402 compatibility is
+        /// needed, the user can force formatting into a string by passing the appropriate
+        /// formatter.
+        ///
+        /// > Note: 
+        /// > - Should there be a convenience method that prints to string specifically?
+        /// > - Do we need `format_into_parts`?
+        fn format<I, J, W>(list: I, formatter: &mut W) -> fmt::Result
```

Comment:
I think you wanted `list: J` :) 
Normally local names don't make a difference, but since there is a `Formatter` and `Write` and `Formatter` implements `Write`, I'd prefer not to name the param `formatter` when it is `Write`.

Reply:
Good points. Done.